File: e2e/tests/implicit.spec.ts
import { test, expect } from '@playwright/test';
import { setup, startAuthorization, idTokenPayload } from './helpers';

test.describe('Implicit flow (id_token token)', () => {
  test('returns tokens in the fragment and binds them with at_hash', async ({ page }) => {
    const s = await setup(page, 'implicit');
    const { nonce } = await startAuthorization(page, s, {
      responseType: 'id_token token',
      nonce: true,
    });

    // The callback page relays the fragment parameters back to the dashboard.
    const result = page.locator('#authorize_result');
    await expect(result).toContainText('"access_token"');
    await expect(result).toContainText('"id_token"');
    await expect(result).not.toContainText('"code"');

    const payload = await idTokenPayload(page);
    expect(payload.iss).toBe('dummy');
    expect(payload.sub).toBe(s.userId);
    expect(payload.aud).toBe(s.app.uid);
    expect(payload.nonce).toBe(nonce);
    expect(payload.at_hash).toBeTruthy();

    // The fragment access token is auto-filled; it must work against userinfo.
    await page.getByRole('button', { name: 'Get UserInfo' }).click();
    await expect(page.locator('#userinfo_result')).toContainText(`"sub": "${s.userId}"`);
  });
});

File: spec/e2e/implicit_spec.rb
# frozen_string_literal: true

require_relative "e2e_helper"

RSpec.describe "Implicit flow (id_token token)" do
  it "returns tokens in the fragment and binds them with at_hash" do
    client = dashboard_client("implicit")
    nonce = start_authorization(client, response_type: "id_token token", nonce: true)

    # The callback page relays the fragment parameters back to the dashboard.
    expect(page).to have_css("#authorize_result", text: '"access_token"')
    expect(page).to have_css("#authorize_result", text: '"id_token"')
    expect(page).to have_no_css("#authorize_result", text: '"code"')

    payload = id_token_payload
    expect(payload["iss"]).to eq("dummy")
    expect(payload["sub"]).to eq(client.user_id)
    expect(payload["aud"]).to eq(client.uid)
    expect(payload["nonce"]).to eq(nonce)
    expect(payload["at_hash"]).to be_present

    # The dashboard fills in the fragment's access token; it must work.
    click_button "Get UserInfo"
    expect(page).to have_css("#userinfo_result", text: %("sub": "#{client.user_id}"))
  end
end
